metals/src/main/scala/scala/meta/internal/metals/debug/SetBreakpointsRequestHandler.scala

```diff
@@ -0,0 +1,86 @@
+package scala.meta.internal.metals.debug
+import org.eclipse.lsp4j.debug.SetBreakpointsArguments
+import org.eclipse.lsp4j.debug.SetBreakpointsResponse
+import org.eclipse.lsp4j.debug.Source
+import org.eclipse.lsp4j.debug.SourceBreakpoint
+import scala.concurrent.ExecutionContext
+import scala.concurrent.Future
+import scala.meta.internal.metals.JvmSignatures
+import scala.meta.internal.metals.MetalsEnrichments._
+import scala.meta.internal.mtags.Mtags
+import scala.meta.internal.semanticdb.SymbolOccurrence
+
+private[debug] final class SetBreakpointsRequestHandler(
+    server: ServerAdapter,
+    adapters: MetalsDebugAdapters
+)(implicit ec: ExecutionContext) {
+
+  // TODO: https://github.com/scalameta/metals/issues/1195
+  def apply(
+      request: SetBreakpointsArguments
+  ): Future[SetBreakpointsResponse] = {
+    val path = adapters.adaptPath(request.getSource.getPath)
+    val input = path.toAbsolutePath.toInput
+    val occurrences = Mtags.allToplevels(input).occurrences
+    val groups = request.getBreakpoints.groupBy { breakpoint =>
```

Comment:
Are we sending fqcns only for symbols from third-party deps? Or are we sending fqcns for symbols for Java and Scala sources defined in the project? Are we following the recommendation in [the PR introducing support for breakpoints in Bloop](https://github.com/scalacenter/bloop/pull/1087)?

Reply:
We are avoiding extra branching by not checking whether the source is defined within the workspace or is a part of a dependency. We are confident this approach is working due to the extensive suite of tests we provide alongside. If you know a case when this will fail, please don't hesitate to point to it.

Or did you have another recommendation in mind?

Reply:
My recommendation is not to send `dap-fqcn` for files defined inside user-defined projects. It should only be used for third-party dependencies.

Bloop and Zinc are the source of truth here. Regardless of how confident we're from Metals doing the right thing, we should be relying on the source of truth if that's possible to avoid issues/discrepancies in the future.

Reply:
Fair enough. For now, though, we will rely on the compiler to emit `LineNumberTable` according to the JVM spec. 
If the compiler, at any point, proves to be unreliable by messing the line numbers, then I believe neither bloop nor zinc would help us.

Reply:
I feel pretty strongly this is not the way to go. You should not be reimplementing this part of the logic. Metals should be reusing the breakpoint logic for user-defined sources implemented in Bloop.

Reply:
The logic you say I am reimplementing is required for the library sources anyway, since bloop solution works only for workspace sources. 
Since it is required either way, right now, I don't see a reason for falling back to bloop and not using it also for user-defined sources. It does not mean there is no such reason but only that I don't see it currently - can you help me in this regard?

Reply:
> The logic you say I am reimplementing is required for the library sources anyway, since bloop solution works only for workspace sources.
  
Yes, the whole reason why we have `dap-fqcn` is so that IDEs that manage library sources can pass in the resolved fully qualified class names.
  
>  I don't see a reason for falling back to bloop and not using it also for user-defined sources. It does not mean there is no such reason but only that I don't see it currently - can you help me in this regard?
  
Yes, the point I'm trying to make is that the logic for library sources is necessary since Bloop doesn't support it. But the logic for workspace sources isn't. There is a reference implementation that it's as good as it gets: it's using the same resolution mechanism the JVM uses for debugging (what's happening under the hood is that the JVM gets our fully qualified class name we pass it and undoes some of the work Bloop has done and finds the class file the loaded class was loaded from in the classloader).
  
At a high level, the process looks like this:
  
Bloop: `$WORKSPACE_SOURCE_FILE` -> `$WORKSPACE_CLASS_FILE` -> `$FQCN` ->
JVM: `$FQCN` -> `$WORKSPACE_CLASS_FILE`
  
The JVM uses the `LineNumberTable` of the `$WORKSPACE_CLASS_FILE` to correctly set the breakpoint. This is why it's so important that `WORKSPACE_CLASS_FILE` matches the one the JVM picks and why the Bloop implementation is the source of the truth -- because it mimics exactly the logic of the JVM.
  
What Metals is doing here might be correct based on the semantic-level information you have. But it's creating a layer of indirection that should be avoided as it could produce issues or inconsistent results in the future. I think it's a good practice to rely on the source of truth if it's possible and in this case it is.